apps/omg_eth/lib/omg_eth/ethereum_height_monitor.ex

```diff
@@ -0,0 +1,204 @@
+    GenServer.start_link(__MODULE__, args, name: __MODULE__)
+  end
+
+  #
+  # GenServer behaviors
+  #
+
+  def init(opts) do
+    _ = Logger.info("Starting Ethereum height monitor.")
+    _ = install_alarm_handler()
+
+    state = %__MODULE__{
+      check_interval_ms: Keyword.fetch!(opts, :check_interval_ms),
+      stall_threshold_ms: Keyword.fetch!(opts, :stall_threshold_ms),
+      synced_at: DateTime.utc_now(),
+      eth_module: Keyword.fetch!(opts, :eth_module),
+      alarm_module: Keyword.fetch!(opts, :alarm_module),
+      event_bus: Keyword.fetch!(opts, :event_bus)
+    }
+
+    {:ok, tref} = :timer.send_after(state.check_interval_ms, :check_new_height)
+    {:ok, %{state | tref: tref}}
+  end
+
+  def handle_info(:check_new_height, state) do
+    height = fetch_height(state.eth_module)
+    stalled? = stalled?(height, state.ethereum_height, state.synced_at, state.stall_threshold_ms)
+
+    :ok = broadcast_on_new_height(state.event_bus, state.ethereum_height, height)
+    _ = connection_alarm(state.alarm_module, state.connection_alarm_raised, height)
+    _ = stall_alarm(state.alarm_module, state.stall_alarm_raised, stalled?)
+
+    state =
+      case height > state.ethereum_height do
+        true -> %{state | ethereum_height: height, synced_at: DateTime.utc_now()}
+        false -> state
+      end
+
+    {:ok, tref} = :timer.send_after(state.check_interval_ms, :check_new_height)
+    {:noreply, %{state | tref: tref}}
+  end
+
+  #
+  # Handle incoming alarms
+  #
+  # These functions are called by the AlarmHandler so that this monitor process can update
+  # its internal state according to the raised alarms. Mainly these handlers do 2 things:
+  #
+  # 1. Reflect the internal state so that it does not re-raise an existing alarm
+  # 2. Maintain the list of events that the alarm corresponds to
+  #
+  def handle_cast({:set_alarm, :ethereum_connection_error}, state) do
+    {:noreply, %{state | connection_alarm_raised: true}}
+  end
+
+  def handle_cast({:clear_alarm, :ethereum_connection_error}, state) do
+    {:noreply, %{state | connection_alarm_raised: false}}
+  end
+
+  def handle_cast({:set_alarm, :ethereum_stalled_sync}, state) do
+    {:noreply, %{state | stall_alarm_raised: true}}
+  end
+
+  def handle_cast({:clear_alarm, :ethereum_stalled_sync}, state) do
+    {:noreply, %{state | stall_alarm_raised: false}}
+  end
+
+  #
+  # Private functions
+  #
+
+  @spec stalled?(non_neg_integer() | :error, non_neg_integer(), DateTime.t(), non_neg_integer()) :: boolean()
+  defp stalled?(height, previous_height, synced_at, stall_threshold_ms) do
+    case height do
+      height when is_integer(height) and height > previous_height ->
+        false
+
+      _ ->
+        DateTime.diff(DateTime.utc_now(), synced_at, :millisecond) > stall_threshold_ms
+    end
+  end
+
+  @spec fetch_height(module()) :: non_neg_integer() | :error
+  defp fetch_height(eth_module) do
+    case eth_module.get_ethereum_height() do
```

Comment:
could `OMG.Eth.get_ethereum_height()` exit with an exception? if yes, have you tested for that?

Reply:
Don't think it raises an exception. Expanded a bit in https://github.com/omisego/elixir-omg/pull/1283#discussion_r376538251